caas/kubernetes/provider/specs/v3_test.go

```diff
@@ -0,0 +1,928 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package specs_test
+
+import (
+	"encoding/base64"
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	admissionregistration "k8s.io/api/admissionregistration/v1beta1"
+	core "k8s.io/api/core/v1"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	k8sspecs "github.com/juju/juju/caas/kubernetes/provider/specs"
+	"github.com/juju/juju/caas/specs"
+	"github.com/juju/juju/testing"
+)
+
+type v3SpecsSuite struct {
+	testing.BaseSuite
+}
+
+var _ = gc.Suite(&v3SpecsSuite{})
+
+var version3Header = `
+version: 3
+`[1:]
+
+func (s *v3SpecsSuite) TestParse(c *gc.C) {
+
+	specStrBase := version3Header + `
+containers:
+  - name: gitlab
+    image: gitlab/latest
+    imagePullPolicy: Always
+    command:
+      - sh
+      - -c
+      - |
+        set -ex
+        echo "do some stuff here for gitlab container"
+    args: ["doIt", "--debug"]
+    workingDir: "/path/to/here"
+    ports:
+      - containerPort: 80
+        name: fred
+        protocol: TCP
+      - containerPort: 443
+        name: mary
+    kubernetes:
+      securityContext:
+        runAsNonRoot: true
+        privileged: true
+      livenessProbe:
+        initialDelaySeconds: 10
+        httpGet:
+          path: /ping
+          port: 8080
+      readinessProbe:
+        initialDelaySeconds: 10
+        httpGet:
+          path: /pingReady
+          port: www
+    config:
```

Comment:
envConfig

Reply:
As discussed,  we will rename this in #11232